### PR TITLE
pkg/ipcache: create a GetK8sMetadata for public access

### LINKED ISSUE #11832: DATA RACE in pkg/ipcache/ipcache.go
```
2020-06-02T15:51:08.16755645Z WARNING: DATA RACE
2020-06-02T15:51:08.167559649Z Write at 0x00c00041ff80 by goroutine 74:
2020-06-02T15:51:08.167562714Z   runtime.mapassign_faststr()
2020-06-02T15:51:08.167565729Z       /usr/local/go/src/runtime/map_faststr.go:202 +0x0
2020-06-02T15:51:08.167568736Z   github.com/cilium/cilium/pkg/ipcache.(*IPCache).Upsert()
2020-06-02T15:51:08.167571773Z       /go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:355 +0xb53
2020-06-02T15:51:08.167574822Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).updatePodHostData()
2020-06-02T15:51:08.167577927Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:495 +0x563
2020-06-02T15:51:08.16758095Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).addK8sPodV1()
2020-06-02T15:51:08.167583957Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:191 +0x4a3
2020-06-02T15:51:08.167586962Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).updateK8sPodV1()
2020-06-02T15:51:08.167590019Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:217 +0x92
2020-06-02T15:51:08.167593043Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).createPodController.func2()
2020-06-02T15:51:08.1675961Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:82 +0x13d
2020-06-02T15:51:08.167599153Z   k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate()
2020-06-02T15:51:08.167602119Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:225 +0xa5
2020-06-02T15:51:08.167605502Z   k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnUpdate()
2020-06-02T15:51:08.16760897Z       <autogenerated>:1 +0x2e
2020-06-02T15:51:08.167612364Z   github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1()
2020-06-02T15:51:08.167615568Z       /go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:113 +0x372
2020-06-02T15:51:08.167618732Z   k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop()
2020-06-02T15:51:08.167621921Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:492 +0x396
2020-06-02T15:51:08.167625278Z   k8s.io/client-go/tools/cache.(*controller).processLoop()
2020-06-02T15:51:08.167640917Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:173 +0x83
2020-06-02T15:51:08.167646241Z   k8s.io/client-go/tools/cache.(*controller).processLoop-fm()
2020-06-02T15:51:08.167649114Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:171 +0x41
2020-06-02T15:51:08.167651935Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
2020-06-02T15:51:08.167656084Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x75
2020-06-02T15:51:08.167659069Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
2020-06-02T15:51:08.167661838Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb3
2020-06-02T15:51:08.167664665Z   k8s.io/apimachinery/pkg/util/wait.JitterUntil()
2020-06-02T15:51:08.167667392Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x10d
2020-06-02T15:51:08.16767018Z   k8s.io/apimachinery/pkg/util/wait.Until()
2020-06-02T15:51:08.167672908Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4a1
2020-06-02T15:51:08.167675696Z   k8s.io/client-go/tools/cache.(*controller).Run()
2020-06-02T15:51:08.167678399Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:145 +0x443
2020-06-02T15:51:08.16768122Z 
2020-06-02T15:51:08.16768389Z Previous read at 0x00c00041ff80 by goroutine 272:
2020-06-02T15:51:08.167686602Z   runtime.mapaccess2_faststr()
2020-06-02T15:51:08.167689265Z       /usr/local/go/src/runtime/map_faststr.go:107 +0x0
2020-06-02T15:51:08.167691996Z   github.com/cilium/cilium/pkg/ipcache.(*IPCache).GetK8sMetadata()
2020-06-02T15:51:08.167694745Z       /go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:166 +0x15a
2020-06-02T15:51:08.167697479Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).GetK8sMetadata()
2020-06-02T15:51:08.167700191Z       /go/src/github.com/cilium/cilium/daemon/cmd/hubble.go:232 +0x51
2020-06-02T15:51:08.167702957Z   github.com/cilium/cilium/pkg/hubble/parser/threefour.(*Parser).resolveEndpoint()
2020-06-02T15:51:08.167706028Z       /go/src/github.com/cilium/cilium/pkg/hubble/parser/threefour/parser.go:266 +0x689
2020-06-02T15:51:08.167709028Z   github.com/cilium/cilium/pkg/hubble/parser/threefour.(*Parser).Decode()
2020-06-02T15:51:08.167711997Z       /go/src/github.com/cilium/cilium/pkg/hubble/parser/threefour/parser.go:154 +0x723
2020-06-02T15:51:08.16771503Z   github.com/cilium/cilium/pkg/hubble/parser.(*Parser).Decode()
2020-06-02T15:51:08.16771799Z       /go/src/github.com/cilium/cilium/pkg/hubble/parser/new.go:73 +0xbd
2020-06-02T15:51:08.167720948Z   github.com/cilium/cilium/pkg/hubble/observer.decodeFlow()
2020-06-02T15:51:08.16772388Z       /go/src/github.com/cilium/cilium/pkg/hubble/observer/local_observer.go:339 +0xa8
2020-06-02T15:51:08.167730007Z   github.com/cilium/cilium/pkg/hubble/observer.(*LocalObserverServer).Start()
2020-06-02T15:51:08.167733174Z       /go/src/github.com/cilium/cilium/pkg/hubble/observer/local_observer.go:155 +0x1c2
2020-06-02T15:51:08.167736298Z 
2020-06-02T15:51:08.167739129Z Goroutine 74 (running) created at:
2020-06-02T15:51:08.167742035Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).podsInit.func1()
2020-06-02T15:51:08.167745435Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:123 +0x399
2020-06-02T15:51:08.167748453Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).podsInit()
2020-06-02T15:51:08.167751383Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:132 +0x7fb
2020-06-02T15:51:08.167754372Z 
2020-06-02T15:51:08.167757172Z Goroutine 272 (running) created at:
2020-06-02T15:51:08.167760072Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).launchHubble()
2020-06-02T15:51:08.167763018Z       /go/src/github.com/cilium/cilium/daemon/cmd/hubble.go:107 +0x774
2020-06-02T15:51:08.167766123Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-06-02T15:51:08.167769213Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1425 +0x1410
2020-06-02T15:51:08.167772628Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2020-06-02T15:51:08.167775956Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:496 +0x2bb6
2020-06-02T15:51:08.167779121Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-06-02T15:51:08.167782096Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1281 +0x354
2020-06-02T15:51:08.167785432Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-06-02T15:51:08.167788706Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:116 +0xab
2020-06-02T15:51:08.16779201Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-06-02T15:51:08.1677953Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:114 +0x91
2020-06-02T15:51:08.167798257Z   github.com/spf13/cobra.(*Command).execute()
2020-06-02T15:51:08.167801022Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:846 +0x8e0
2020-06-02T15:51:08.167803856Z   github.com/spf13/cobra.(*Command).ExecuteC()
2020-06-02T15:51:08.167806563Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:950 +0x499
2020-06-02T15:51:08.167809372Z   github.com/spf13/cobra.(*Command).Execute()
2020-06-02T15:51:08.167812032Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:887 +0x1eb
2020-06-02T15:51:08.167814808Z   github.com/cilium/cilium/daemon/cmd.Execute()
2020-06-02T15:51:08.167817509Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:147 +0x1cc
2020-06-02T15:51:08.16782029Z   main.main()
2020-06-02T15:51:08.167823237Z       /go/src/github.com/cilium/cilium/daemon/main.go:22 +0x2f

```

### PR DESCRIPTION
**Pull request by aanm**
GetK8sMetadata was supposed to be used for internal usage of the
package, making this function public requires protection against
parallel access by its internal mutex.

Fixes: 4699840b0e09 ("ipcache: Export GetK8sMetadata")
Signed-off-by: André Martins <andre@cilium.io>

Fixes https://github.com/cilium/cilium/issues/11832

### CODE REVIEW

**Review by michi-covalent — APPROVED** [review]
ah good to know. thank you for the fix

File: pkg/ipcache/ipcache.go [hunk 1]
```diff
@@ -163,6 +163,16 @@ func (ipc *IPCache) getHostIPCache(ip string) (net.IP, uint8) {
 
 // GetK8sMetadata returns Kubernetes metadata for the given IP address.
 func (ipc *IPCache) GetK8sMetadata(ip string) *K8sMetadata {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	if k8sMeta, ok := ipc.ipToK8sMetadata[ip]; ok {
+		return &k8sMeta
```
Comment (joestringer) [inline pkg/ipcache/ipcache.go]: Should we be returning a pointer to state covered by this mutex? Seems like `updateNamedPorts()` could modify it.

**Review by joestringer — CHANGES_REQUESTED** [review]
Makes sense, one concern below.

File: pkg/ipcache/ipcache.go (hunk 1 above)
Comment (aanm) [reply pkg/ipcache/ipcache.go]: @joestringer as far as I can see, the mutex is protecting the map, not its values. However I've double checked and it seems the value is never modified but I've made it clear in the function documentation.

**Comment by aanm** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
hit https://github.com/cilium/cilium/issues/11847

**Pull request merged**